modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs

```diff
@@ -273,36 +480,207 @@ public void Add(Variant item)
         }
 
         /// <summary>
-        /// Checks if this <see cref="Array"/> contains the given item.
+        /// Adds the elements of the specified collection to the end of this <see cref="Array"/>.
         /// </summary>
+        /// <param name="collection">Collection of <see cref="Variant"/> items to add.</param>
+        /// <exception cref="ArgumentNullException">
+        /// The <paramref name="collection"/> is <see langword="null"/>.
+        /// </exception>
+        public void AddRange(IEnumerable<Variant> collection)
+        {
+            if (collection == null)
+                throw new ArgumentNullException(nameof(collection), "Value cannot be null.");
+
+            // If the collection is another Godot Array, we can add the items
+            // with a single interop call.
+            if (collection is Array array)
+            {
+                var self = (godot_array)NativeValue;
+                var collectionNative = (godot_array)array.NativeValue;
+                _ = NativeFuncs.godotsharp_array_add_range(ref self, collectionNative);
+                return;
+            }
+            if (collection is IGenericGodotArray typedArray)
+            {
+                var self = (godot_array)NativeValue;
+                var collectionNative = (godot_array)typedArray.UnderlyingArray.NativeValue;
+                _ = NativeFuncs.godotsharp_array_add_range(ref self, collectionNative);
+                return;
+            }
+
+            // If we can retrieve the count of the collection without enumerating it
+            // (e.g.: the collections is a List<T>), use it to resize the array once
+            // instead of growing it as we add items.
+            if (collection.TryGetNonEnumeratedCount(out int count))
+            {
+                Resize(Count + count);
+
+                var enumerator = collection.GetEnumerator();
+
+                for (int i = 0; i < count; i++)
+                {
+                    enumerator.MoveNext();
+                    this[count + i] = enumerator.Current;
+                }
+
+                return;
+            }
+
+            foreach (var item in collection)
+            {
+                Add(item);
+            }
+        }
+
+        /// <summary>
+        /// Finds the index of an existing value (or the insertion index that
+        /// maintains sorting order, if the value is not yet present in the array)
```

Comment:
While I argued for the inclusion of this method, this is not how binary search works in .net: If the value does not exist in .net the complement if the insertion index is returned so that you can directly check if the element exists without having to do another lookup in the array. And the weird before argument just doesn't exist.

Reply:
I see, for me the main reason to add the method is the search can be done in C++ avoiding marshaling every item which would be very expensive. I don't particularly care about which implementation is used although I'm more inclined to follow the existing .NET precedent.

Reply:
I've now implemented `BinarySearch` the same way it works in .NET's `Array.BinarySearch`.